File: backend/importar/extrair_prazos_pad_sinve.php
<?php
// ============================================================================
//  Extração de prazos PAD/SINVE (alta confiança) — espelho PHP de
//  extrair_prazos_pad_sinve.py. Rodado durante importação de cada ato.
//
//  Classificação:
//    - Tipo: PAD, PAD_SUMARIO, SINVE, SINDACUS
//    - Papel: INSTAURACAO, EXTENSAO, SOBRESTAMENTO
//    - Base legal: dias literais extraídos do texto
//
//  Uso (CLI):
//      php -r "require 'extrair_prazos_pad_sinve.php'; var_dump(extrair_prazos_pad_sinve(...));"
//
//  Uso (no importador):
//      foreach (extrair_prazos_pad_sinve($ementaP, $textoP, $dataP) as $pz) { ... }
// ============================================================================

function extrair_prazos_pad_sinve(string $ementa, string $texto, ?string $data_ato): array {
    $blob = mb_strtolower($ementa . " " . $texto, 'UTF-8');

    // Classificação por tipo. As SINDICÂNCIAS vêm antes do PAD: são o termo
    // mais específico, e uma sindicância costuma citar "processo administrativo
    // disciplinar" como possível desdobramento — checar PAD antes rotularia
    // esses atos como PAD por engano.
    // (Espelha classifica_tipo em extrair_prazos_pad_sinve.py.)
    //
    // Sindicância ACUSATÓRIA (também chamada punitiva) e INVESTIGATIVA são
    // instrumentos diferentes: a investigativa é inquisitorial e não pune; a
    // acusatória tem contraditório e pode aplicar penalidade. Misturar as duas
    // numa categoria só apaga a distinção que interessa a quem consulta.
    //
    // Um mesmo ato pode citar as duas ("converter a investigativa em
    // acusatória"), então vence a que aparece PRIMEIRO — e o blob começa pela
    // ementa, que é onde o ato declara o que ele próprio faz.
    $posAcus = preg_match('/sindic[âa]ncia\s+(?:acusat[óo]ria|punitiva)/u', $blob, $mA, PREG_OFFSET_CAPTURE)
             ? $mA[0][1] : PHP_INT_MAX;
    $posInv  = preg_match('/sindic[âa]ncia\s+investigat[óo]ria|sindic[âa]ncia\s+investigativa/u', $blob, $mI, PREG_OFFSET_CAPTURE)
             ? $mI[0][1] : PHP_INT_MAX;

    if ($posAcus < $posInv) {
        $tipo = 'SINDACUS';
    } elseif ($posInv !== PHP_INT_MAX) {
        $tipo = 'SINVE';
    } elseif (preg_match('/processo\s+administrativo\s+disciplinar\s+sum[aá]rio/u', $blob)) {
        $tipo = 'PAD_SUMARIO';
    } elseif (preg_match('/processo\s+administrativo\s+disciplinar/u', $blob)) {
        $tipo = 'PAD';
    } else {
        return [];  // fora do escopo PAD/SINVE
    }

    // Classificação por papel (só primeiros ~600 chars pra evitar falsos positivos)
    $inicio = mb_substr($blob, 0, 600);
    if (preg_match('/\breconduz\w*\b|\bprorroga\w*\b/', $inicio)) {
        $papel = 'EXTENSAO';
    } elseif (preg_match('/\binstaura(r|ção|da|do|m)?\b/', $inicio)) {
        $papel = 'INSTAURACAO';
    } elseif (preg_match('/\bsobrest\w*\b/', $inicio)) {
        $papel = 'SOBRESTAMENTO';
    } else {
        return [];  // papel desconhecido
    }

    // Extração de dias: "prazo de 30 (trinta) dias" ou "prorrogar por 60 dias"
    $dias = null;
    if (preg_match('/prazo\s+(?:\S+\s+){0,2}?de\s+(\d{1,3})\s*\([^)]{0,25}\)?\s*dias/iu', $blob, $m)) {
        $dias = (int)$m[1];
    } elseif (preg_match('/prorroga\w*\s+(?:o\s+prazo\s+)?por\s+(?:mais\s+)?(\d{1,3})\s*\(?[^)]{0,25}\)?\s*dias/iu', $blob, $m)) {
        $dias = (int)$m[1];
    }

    if ($dias === null) {
        return [];  // sem dias literal declarado
    }

    // Calcular data_limite: data_ato + dias (regra lei 9784: exclui dia inicial)
    // Simplificação: usar dias corridos (não úteis), sem feriado lookup
    if ($data_ato && preg_match('/^\d{4}-\d{2}-\d{2}$/', $data_ato)) {
        $dt = DateTime::createFromFormat('Y-m-d', $data_ato);
        if ($dt) {
            $dt->modify("+{$dias} days");
            $data_limite = $dt->format('Y-m-d');
        } else {
            $data_limite = null;
        }
    } else {
        $data_limite = null;
    }

    // Rótulo "público" determinístico (mesmo no backfill e no import diário)
    $publico_map = [
        'PAD'         => 'Comissão de PAD',
        'PAD_SUMARIO' => 'Comissão de PAD Sumário',
        'SINVE'       => 'Comissão de Sindicância',
        'SINDACUS'    => 'Comissão de Sindicância',
    ];
    $publico = $publico_map[$tipo] ?? 'Comissão';

    $papel_label = [
        'INSTAURACAO'   => 'instauração',
        'EXTENSAO'      => 'prorrogação/recondução',
        'SOBRESTAMENTO' => 'sobrestamento',
    ][$papel] ?? $papel;
    $tipo_label = [
        'PAD'         => 'PAD',
        'PAD_SUMARIO' => 'PAD Sumário',
        'SINVE'       => 'Sindicância Investigativa',
        'SINDACUS'    => 'Sindicância Acusatória',
    ][$tipo] ?? $tipo;

    // Resultado: um "prazo" de alta confiança
    return [[
        'tipo' => $tipo,           // PAD, PAD_SUMARIO, SINVE, SINDACUS (código; o front rotula)
        'papel' => $papel,         // INSTAURACAO, EXTENSAO, SOBRESTAMENTO
        'dias' => $dias,
        'dataLimite' => $data_limite,
        'conf' => 'alta',   // minúsculo: casa com o filtro do front (p.conf==='alta')
        'base' => 'PAD_SINVE',
        'publico' => $publico,
        'origem' => "{$tipo_label} · {$papel_label} · prazo de {$dias} dias",
    ]];
}

File: backend/importar/extrair_prazos_pad_sinve.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Extração de prazos PAD/SINVE (Processo Administrativo Disciplinar &
Sindicância Investigativa) com alta confiança. Diferente dos prazos
genéricos, estes são sempre literais (30, 60, 45, 15 dias) e rastreáveis
a uma cadeia de instauração → prorrogações → sobrestamentos.

Uso (validação local contra dump):
    python extrair_prazos_pad_sinve.py < pad_sinve.json > prazos_pad_sinve.json

Uso em produção (ser chamado pelo importar_v2.php):
    php -r "include 'extrair_prazos_pad_sinve.php'; ..."
"""

import json
import re
import sys
from collections import defaultdict

def norm(s):
    return (s or "").lower()

# Classificação por tipo de procedimento
PAD_SUM_RE = re.compile(r"processo administrativo disciplinar\s+sum[aá]rio", re.I)
PAD_RE = re.compile(r"processo administrativo disciplinar", re.I)
SINVE_RE = re.compile(r"sindic[âa]ncia investigat[óo]ria|sindic[âa]ncia investigativa", re.I)

# Classificação por papel no fluxo
INSTAURA_RE = re.compile(r"\binstaura(r|ção|da|do|m)?\b", re.I)
RECONDUZ_RE = re.compile(r"\breconduz\w*\b", re.I)
PRORROGA_RE = re.compile(r"\bprorroga\w*\b", re.I)
SOBRESTA_RE = re.compile(r"\bsobrest\w*\b", re.I)

# Extração de dias declarados: "prazo de 30 (trinta) dias" ou "prorrogar por mais 60 (sessenta) dias"
PRAZO_DIAS_RE = re.compile(
    r"prazo\s+(?:\S+\s+){0,2}?de\s+(\d{1,3})\s*\([^)]{0,25}\)?\s*dias",
    re.I
)
PRORROGA_POR_RE = re.compile(
    r"prorroga\w*\s+(?:o\s+prazo\s+)?por\s+(?:mais\s+)?(\d{1,3})\s*\(?[^)]{0,25}\)?\s*dias",
    re.I
)

# Citação ao ato originário (instaurador)
CITA_ORIGINARIO_RE = re.compile(
    r"(?:instaurad[oa]|institu[íi]d[oa]|designad[oa]|constitu[íi]d[oa])\s+"
    r"(?:atrav[ée]s\s+d[ae]|pel[ao]|n[ao])?\s*"
    r"(portaria|dts|determina[çc][ãa]o\s+de\s+servi[çc]o)\s*(?:n[ºo°.]*\s*)?\s*([\d.]+)"
    r"(?:\s*[-/,]?\s*(?:de\s*)?(\d{1,2}[./]\d{1,2}[./]\d{2,4}|\d{1,2}\s+de\s+\w+\s+de\s+\d{4}))?",
    re.I
)

def classifica_tipo(txt):
    """Retorna 'SINVE', 'PAD_SUMARIO', 'PAD', ou None.

    SINVE é checado PRIMEIRO: 'sindicância investigativa' é o termo mais
    específico e distintivo. Uma SINVE costuma citar 'processo administrativo
    disciplinar' como possível desdobramento — se checássemos PAD antes, esses
    atos seriam rotulados PAD por engano.
    """
    if SINVE_RE.search(txt):
        return "SINVE"
    if PAD_SUM_RE.search(txt):
        return "PAD_SUMARIO"
    if PAD_RE.search(txt):
        return "PAD"
    return None

def classifica_papel(txt):
    """Retorna 'INSTAURACAO', 'EXTENSAO', ou 'OUTRO'."""
    inicio = txt[:600]
    if SOBRESTA_RE.search(inicio):
        return "SOBRESTAMENTO"
    if RECONDUZ_RE.search(inicio) or PRORROGA_RE.search(inicio):
        return "EXTENSAO"
    if INSTAURA_RE.search(inicio):
        return "INSTAURACAO"
    return "OUTRO"

def extrai_dias(txt):
    """Retorna nº de dias declarado, ou None."""
    m = PRAZO_DIAS_RE.search(txt) or PRORROGA_POR_RE.search(txt)
    return int(m.group(1)) if m else None

def extrai_citacao_originario(txt):
    """Retorna dict com tipo, numero, data da portaria originária, ou None."""
    m = CITA_ORIGINARIO_RE.search(txt)
    if not m:
        return None
    tipo_cit = m.group(1)
    numero_cit = m.group(2)
    # Deixar data por enquanto, mas não vamos usar pra matching (chave é tipo+numero+ano do ato)
    return {"tipo": tipo_cit.strip().upper(), "numero": numero_cit}

def processa_atos(atos_json_list):
    """
    Processa lista de atos (do dump ou JSON feed), retorna lista de prazos PAD/SINVE.
    Cada prazo tem:
      - uid, tipo_ato (PAD/SINVE), papel (INSTAURACAO/EXTENSAO/SOBRESTAMENTO)
      - dias (nº literal extraído do texto)
      - processo_sei (se houver)
      - ato_originario_uid (uid do ato que instaurou, se prorrogação)
      - vigente (bool: True se é o ato mais recente da cadeia process_sei)
    """
    prazos = []
    por_sei = defaultdict(list)

    for a in atos_json_list:
        txt_full = norm(a.get("ementa", "")) + " " + norm(a.get("texto_original", ""))
        tipo = classifica_tipo(txt_full)
        papel = classifica_papel(txt_full)

        if tipo is None or papel == "OUTRO":
            continue  # fora do escopo PAD/SINVE

        dias = extrai_dias(txt_full)
        if dias is None:
            continue  # sem prazo literal declarado

        sei = a.get("processo_sei")
        data_ato = a.get("data_ato")

        publico = {
            "PAD": "Comissão de PAD",
            "PAD_SUMARIO": "Comissão de PAD Sumário",
            "SINVE": "Comissão de Sindicância",
        }.get(tipo, "Comissão")
        papel_label = {
            "INSTAURACAO": "instauração",
            "EXTENSAO": "prorrogação/recondução",
            "SOBRESTAMENTO": "sobrestamento",
        }.get(papel, papel)

        p = {
            "uid": a["uid"],
            "tipo_ato": tipo,
            "papel": papel,
            "dias": dias,
            "processo_sei": sei,
            "data_ato": data_ato,
            "publico": publico,
            "origem": f"{tipo} · {papel_label} · prazo de {dias} dias",
        }

        # Se é extensão, tenta extrair citação ao originário
        if papel == "EXTENSAO":
            cit = extrai_citacao_originario(txt_full)
            if cit:
                p["citacao_tipo"] = cit["tipo"]
                p["citacao_numero"] = cit["numero"]

        prazos.append(p)

        # Agrupar por processo_sei pra marcar qual é vigente depois
        if sei:
            por_sei[sei].append((data_ato or "", p))

    # Marcar vigente: o de maior data_ato dentro de cada processo_sei
    vigentes_por_sei = set()
    for sei, lista in por_sei.items():
        lista.sort(key=lambda x: x[0], reverse=True)  # sort por data_ato (strings AAAA-MM-DD sortem lexicograficamente)
        if lista:
            vigentes_por_sei.add(lista[0][1]["uid"])

    for p in prazos:
        p["vigente"] = p["uid"] in vigentes_por_sei

    return prazos

if __name__ == "__main__":
    # Lê JSON do stdin (dump_pad_sinve.json)
    dados = json.load(sys.stdin)
    prazos = processa_atos(dados)
    json.dump(prazos, sys.stdout, ensure_ascii=False, indent=2)
